File: backend/app/routers/factory.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import Optional
from app.database import get_db
from app.models.factory import FactoryLocation, Workstation, StorageLocation, WorkstationCategory
from app.auth import require_admin, get_current_user

router = APIRouter(prefix="/api/factory", tags=["factory"])


# ── Factory Locations ──────────────────────────────────────────────────────────

@router.get("/locations")
def list_locations(db: Session = Depends(get_db), _=Depends(get_current_user)):
    locs = db.query(FactoryLocation).filter(FactoryLocation.is_active == True).all()
    return [{"id": l.id, "code": l.code, "name": l.name} for l in locs]


class LocationCreate(BaseModel):
    code: str
    name: str


class LocationUpdate(BaseModel):
    name: Optional[str] = None
    code: Optional[str] = None


@router.patch("/locations/{loc_id}")
def update_location(loc_id: int, body: LocationUpdate, db: Session = Depends(get_db), _=Depends(require_admin)):
    loc = db.query(FactoryLocation).filter(FactoryLocation.id == loc_id).first()
    if not loc:
        raise HTTPException(status_code=404, detail="Location not found")
    for field, value in body.model_dump(exclude_none=True).items():
        setattr(loc, field, value)
    db.commit()
    db.refresh(loc)
    return {"id": loc.id, "code": loc.code, "name": loc.name}


@router.post("/locations", status_code=201)
def create_location(body: LocationCreate, db: Session = Depends(get_db), _=Depends(require_admin)):
    loc = FactoryLocation(code=body.code, name=body.name)
    db.add(loc)
    db.commit()
    db.refresh(loc)
    return {"id": loc.id, "code": loc.code, "name": loc.name}


# ── Workstations ───────────────────────────────────────────────────────────────

def ws_out(w: Workstation) -> dict:
    return {
        "id": w.id, "code": w.code, "name": w.name,
        "category": w.category, "is_active": w.is_active,
        "factory_location_id": w.factory_location_id,
    }


@router.get("/workstations")
def list_workstations(location_id: Optional[int] = None, db: Session = Depends(get_db), _=Depends(get_current_user)):
    q = db.query(Workstation).filter(Workstation.is_active == True)
    if location_id:
        q = q.filter(
            (Workstation.factory_location_id == location_id) | (Workstation.factory_location_id == None)
        )
    return [ws_out(w) for w in q.all()]


class WorkstationCreate(BaseModel):
    code: str
    name: str
    category: WorkstationCategory = WorkstationCategory.other
    factory_location_id: Optional[int] = None


class WorkstationUpdate(BaseModel):
    name: Optional[str] = None
    category: Optional[WorkstationCategory] = None
    factory_location_id: Optional[int] = None
    is_active: Optional[bool] = None


@router.post("/workstations", status_code=201)
def create_workstation(body: WorkstationCreate, db: Session = Depends(get_db), _=Depends(require_admin)):
    ws = Workstation(**body.model_dump())
    db.add(ws)
    db.commit()
    db.refresh(ws)
    return ws_out(ws)


@router.patch("/workstations/{ws_id}")
def update_workstation(ws_id: int, body: WorkstationUpdate, db: Session = Depends(get_db), _=Depends(require_admin)):
    ws = db.query(Workstation).filter(Workstation.id == ws_id).first()
    if not ws:
        raise HTTPException(status_code=404, detail="Workstation not found")
    for field, value in body.model_dump(exclude_none=True).items():
        setattr(ws, field, value)
    db.commit()
    db.refresh(ws)
    return ws_out(ws)


# ── Storage Locations ──────────────────────────────────────────────────────────

@router.get("/storage")
def list_storage(db: Session = Depends(get_db), _=Depends(get_current_user)):
    locs = db.query(StorageLocation).filter(StorageLocation.is_active == True).all()
    return [{"id": l.id, "code": l.code, "name": l.name, "factory_location_id": l.factory_location_id} for l in locs]


class StorageCreate(BaseModel):
    code: str
    name: Optional[str] = None
    factory_location_id: Optional[int] = None


@router.post("/storage", status_code=201)
def create_storage(body: StorageCreate, db: Session = Depends(get_db), _=Depends(require_admin)):
    s = StorageLocation(**body.model_dump())
    db.add(s)
    db.commit()
    db.refresh(s)
    return {"id": s.id, "code": s.code, "name": s.name, "factory_location_id": s.factory_location_id}

File: backend/src/routes/factory.js
import { Router } from 'express';
import { query, one } from '../db/pool.js';
import { asyncHandler } from '../middleware/error.js';
import { requireAuth, requireAdmin, HttpError } from '../middleware/auth.js';

const router = Router();

function wsOut(w) {
  return {
    id: w.id,
    code: w.code,
    name: w.name,
    category: w.category,
    is_active: w.is_active,
    factory_location_id: w.factory_location_id,
  };
}

// ── Factory Locations ──────────────────────────────────────────────────────
router.get(
  '/locations',
  requireAuth,
  asyncHandler(async (req, res) => {
    const rows = await query('SELECT * FROM factory_locations WHERE is_active = TRUE ORDER BY id');
    res.json(rows.map((l) => ({ id: l.id, code: l.code, name: l.name })));
  })
);

router.post(
  '/locations',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const { code, name } = req.body || {};
    const l = await one('INSERT INTO factory_locations (code, name) VALUES ($1,$2) RETURNING *', [code, name]);
    res.status(201).json({ id: l.id, code: l.code, name: l.name });
  })
);

router.patch(
  '/locations/:locId',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = parseInt(req.params.locId, 10);
    const loc = await one('SELECT * FROM factory_locations WHERE id = $1', [id]);
    if (!loc) throw new HttpError(404, 'Location not found');
    const sets = [];
    const vals = [];
    let i = 1;
    for (const f of ['name', 'code']) {
      if (req.body[f] !== undefined && req.body[f] !== null) {
        sets.push(`${f} = $${i++}`);
        vals.push(req.body[f]);
      }
    }
    if (sets.length) {
      vals.push(id);
      await query(`UPDATE factory_locations SET ${sets.join(', ')} WHERE id = $${i}`, vals);
    }
    const l = await one('SELECT * FROM factory_locations WHERE id = $1', [id]);
    res.json({ id: l.id, code: l.code, name: l.name });
  })
);

// ── Workstations ───────────────────────────────────────────────────────────
router.get(
  '/workstations',
  requireAuth,
  asyncHandler(async (req, res) => {
    const locationId = req.query.location_id ? parseInt(req.query.location_id, 10) : null;
    let rows;
    if (locationId) {
      rows = await query(
        'SELECT * FROM workstations WHERE is_active = TRUE AND (factory_location_id = $1 OR factory_location_id IS NULL) ORDER BY id',
        [locationId]
      );
    } else {
      rows = await query('SELECT * FROM workstations WHERE is_active = TRUE ORDER BY id');
    }
    res.json(rows.map(wsOut));
  })
);

router.post(
  '/workstations',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const { code, name, category = 'Other', factory_location_id = null } = req.body || {};
    const w = await one(
      'INSERT INTO workstations (code, name, category, factory_location_id) VALUES ($1,$2,$3,$4) RETURNING *',
      [code, name, category, factory_location_id]
    );
    res.status(201).json(wsOut(w));
  })
);

router.patch(
  '/workstations/:wsId',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const id = parseInt(req.params.wsId, 10);
    const ws = await one('SELECT * FROM workstations WHERE id = $1', [id]);
    if (!ws) throw new HttpError(404, 'Workstation not found');
    const sets = [];
    const vals = [];
    let i = 1;
    for (const f of ['name', 'category', 'factory_location_id', 'is_active']) {
      if (req.body[f] !== undefined && req.body[f] !== null) {
        sets.push(`${f} = $${i++}`);
        vals.push(req.body[f]);
      }
    }
    if (sets.length) {
      vals.push(id);
      await query(`UPDATE workstations SET ${sets.join(', ')} WHERE id = $${i}`, vals);
    }
    const w = await one('SELECT * FROM workstations WHERE id = $1', [id]);
    res.json(wsOut(w));
  })
);

// ── Storage Locations ──────────────────────────────────────────────────────
router.get(
  '/storage',
  requireAuth,
  asyncHandler(async (req, res) => {
    const rows = await query('SELECT * FROM storage_locations WHERE is_active = TRUE ORDER BY id');
    res.json(rows.map((l) => ({ id: l.id, code: l.code, name: l.name, factory_location_id: l.factory_location_id })));
  })
);

router.post(
  '/storage',
  requireAdmin,
  asyncHandler(async (req, res) => {
    const { code, name = null, factory_location_id = null } = req.body || {};
    const s = await one(
      'INSERT INTO storage_locations (code, name, factory_location_id) VALUES ($1,$2,$3) RETURNING *',
      [code, name, factory_location_id]
    );
    res.status(201).json({ id: s.id, code: s.code, name: s.name, factory_location_id: s.factory_location_id });
  })
);

export default router;
